eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxow.hs

```diff
@@ -312,10 +312,15 @@ babbageUtxowTransition = do
   {-  witsVKeyNeeded utxo tx genDelegs ⊆ witsKeyHashes                   -}
   runTest $ validateNeededWitnesses @era witsKeyHashes certState utxo txBody
   -- TODO can we add the required signers to witsVKeyNeeded so we dont need the check below?
+  --
+  --      ^ Answer: yes we can. It is now implememnted. We just can't remove
+  --      MissingRequiredSigners predicate failure until we are in Conway
 
   {-  THIS DOES NOT APPPEAR IN THE SPEC as a separate check, but
       witsVKeyNeeded must include the reqSignerHashes in the union   -}
   {- reqSignerHashes txbody ⊆ witsKeyHashes -}
+  -- TODO: Remove once we are in Conway. This check is now handled by
+  -- validateNeededWitnesses, as per spec
```

Comment:
Won't it fail with a different Predicate failure in  `validateNeededWitnesses`  (`MissingVKeyWitnessesUTXOW`)  ?  Is that not something we need to be careful about changing?

Reply:
That's a very good question.

There are two important points in ledger that we need to consider with respect to predicate failures:

* If a transaction was invalid for any reason, it should continue being invalid in the future. Regardless of the type of predicate failure produced. That is necessary for consensus. In other words consensus does not care why transaction is invalid all it cares about is that it is invalid, because it doesn't propagate invalid transactions.
* We cannot start producing new types of predicate failures for current era that were not available before without a protocol version bump, because that would break NodeToClient communication protocol on the serialization level.

Neither one of those restrictions are preventing us from producing an extra predicate failure, which is what happening right now in this PR. In the matter of fact, writing this explanation up makes me realize that we can even produce a different predicate failure for the same predicate check, as long as that new type of predicate check existed before. So, technically we can take care of this TODO even now.

What we cannot do until we are in Conway, is to remove the actual `MissingRequiredSigners` type constructor, because that would break NodeToClient communication on the client side. This is because a newer client can be talking to an older node, that is capable producing this predicate failure.